packages/dev/addons/src/navigation/plugin/RecastJSCrowd.ts

```diff
@@ -0,0 +1,439 @@
+     */
+    private _agentDestination: Vector3[] = new Array<Vector3>();
+    /**
+     * Link to the scene is kept to unregister the crowd from the scene
+     */
+    private _scene: Scene;
+
+    private _engine: AbstractEngine;
+
+    /**
+     * Observer for crowd updates
+     */
+    private _onBeforeAnimationsObserver: Nullable<Observer<Scene>> = null;
+
+    /**
+     * Fires each time an agent is in reach radius of its destination
+     */
+    public onReachTargetObservable = new Observable<{
+        /**
+         *
+         */
+        agentIndex: number;
+        /**
+         *
+         */
+        destination: Vector3;
+    }>();
+
+    /**
+     * Constructor
+     * @param plugin recastJS plugin
+     * @param maxAgents the maximum agent count in the crowd
+     * @param maxAgentRadius the maximum radius an agent can have
+     * @param scene to attach the crowd to
+     * @returns the crowd you can add agents to
+     */
+    public constructor(plugin: RecastNavigationJSPluginV2, maxAgents: number, maxAgentRadius: number, scene: Scene) {
+        this.navigationPlugin = plugin;
+
+        if (!plugin.navMesh) {
+            throw new Error("There is no NavMesh generated.");
+        }
+
+        this.recastCrowd = new BjsRecast.Crowd(plugin.navMesh, {
+            maxAgents,
+            maxAgentRadius,
+        });
+
+        this._scene = scene;
+        this._engine = scene.getEngine();
+
+        this._onBeforeAnimationsObserver = scene.onBeforeAnimationsObservable.add(() => {
+            this.update(this._engine.getDeltaTime() * 0.001 * plugin.timeFactor);
+        });
+    }
+
+    /**
+     * Add a new agent to the crowd with the specified parameter a corresponding transformNode.
+     * You can attach anything to that node. The node position is updated in the scene update tick.
+     * @param pos world position that will be constrained by the navigation mesh
+     * @param parameters agent parameters
+     * @param transform hooked to the agent that will be update by the scene
+     * @returns agent index
+     */
+    public addAgent(pos: IVector3Like, parameters: IAgentParametersV2, transform: TransformNode): number {
+        const agentParams = ToCrowdAgentParams(parameters);
+
+        const agent = this.recastCrowd.addAgent({ x: pos.x, y: pos.y, z: pos.z }, agentParams);
+
+        this.transforms.push(transform);
+        this.agents.push(agent.agentIndex);
+        this.reachRadii.push(parameters.reachRadius ? parameters.reachRadius : parameters.radius);
+        this._agentDestinationArmed.push(false);
+        this._agentDestination.push(new Vector3(0, 0, 0));
+
+        return agent.agentIndex;
+    }
+
+    /**
+     * Returns the agent position in world space
+     * @param index agent index returned by addAgent
+     * @returns world space position
+     */
+    public getAgentPosition(index: number): Vector3 {
+        const agentPos = this.recastCrowd.getAgent(index)?.position() ?? {
```

Comment:
You could probably also use `Vector3.ZeroReadonly` here if youw ant to avoid the allocations (not sure how frequently these functions are called).

Reply:
So you mean: ` const agentPos = this.recastCrowd.getAgent(index)?.position() ?? Vector3.ZeroReadOnly`, hm?
The second branch will be rarely hit so this will be a micro optimization only. But why not? :-) Changed. Thanks!